it/lessons/advanced/concurrency.md

```diff
@@ -0,0 +1,175 @@
+    end
+  end
+end
+
+iex> pid = spawn(Example, :listen, [])
+#PID<0.108.0>
+
+iex> send pid, {:ok, "hello"}
+World
+{:ok, "hello"}
+
+iex> send pid, :ok
+:ok
+```
+
+### Process Linking
+
+Un problema con `spawn` è sapere quando il processo lanciato muore. Per questo motivo, abbiamo bisogno di collegare i nostri processi, usando `spawn_link`. Due processi collegati riceveranno notifiche d'uscita vicendevolmente.
+
+```elixir
+defmodule Example do
+  def explode, do: exit(:kaboom)
+end
+
+iex> spawn(Example, :explode, [])
+#PID<0.66.0>
+
+iex> spawn_link(Example, :explode, [])
+** (EXIT from #PID<0.57.0>) :kaboom
+```
+
+A volte, non vogliamo che un processo collegato uccida il processo che l'ha creato. Per evitare questo, dobbiamo _intrappolare le uscite_ nel processo genitore. Quando intrappoliamo uscite in un processo, questo riceverà solo messaggi di notifica in questa forma: `{:EXIT, dal_pid, ragione}`.
+
+```elixir
+defmodule Example do
+  def explode, do: exit(:kaboom)
+  def run do
+    Process.flag(:trap_exit, true)
+    spawn_link(Example, :explode, [])
+
+    receive do
+      {:EXIT, from_pid, reason} -> IO.puts "Exit reason: #{reason}"
+    end
+  end
+end
+
+iex> Example.run
+Exit reason: kaboom
+:ok
+```
+
+### Process Monitoring
+
+Cosa facciamo quando non vogliamo che due processi siano collegati, ma rimangano comunque informati sull'altro? In questo caso, possiamo fare _process monitoring_ con `spawn_monitor`. Quando monitoriamo un processo, riceviamo un messaggio quando questo muore, senza preoccuparci di ripercussioni sul processo corrente, e senza dover intrappolare uscite.
+
+```elixir
+defmodule Example do
+  def explode, do: exit(:kaboom)
+  def run do
+    {pid, ref} = spawn_monitor(Example, :explode, [])
+
+    receive do
+      {:DOWN, ref, :process, from_pid, reason} -> IO.puts "Exit reason: #{reason}"
+    end
+  end
+end
+
+iex> Example.run
+Exit reason: kaboom
+:ok
+```
+
+## Agenti
+
+Gli Agenti sono astrazioni attorno a dei processi di background, che mantengono uno stato interno. Possiamo accedere ad agenti da altri processi all'interno della nostra applicazione (e nodo). Lo stato di un agente è determinato dal valore di ritorno della funzione che lo inizializza, o aggiorna:
+
+```elixir
+iex> {:ok, agent} = Agent.start_link(fn -> [1, 2, 3] end)
+{:ok, #PID<0.65.0>}
+
+iex> Agent.update(agent, fn (state) -> state ++ [4, 5] end)
+:ok
+
+iex> Agent.get(agent, &(&1))
+[1, 2, 3, 4, 5]
+```
+
+Dando un nome ad un agente, possiamo comunicare con esso senza sapere il suo PID:
+
+```elixir
+iex> Agent.start_link(fn -> [1, 2, 3] end, name: Numbers)
+{:ok, #PID<0.74.0>}
+
+iex> Agent.get(Numbers, &(&1))
+[1, 2, 3]
+```
+
+## Task
+
+I Task rendono possibile eseguire una funzione in background, della quale possiamo ottenere il valore di ritorno in futuro. I Task possono essere particolarmente utili quando abbiamo a che fare con operazioni costose, e vogliamo evitare di bloccare l'esecuzione della nostra applicazione.
```

Comment:
I Task rendono possibile eseguire una funzione in background, dalla quale possiamo ottenere il valore di ritorno in un momento successivo. I Task possono essere particolarmente utili quando abbiamo a che fare con operazioni particolarmente onerose in termini di tempo di esecuzione(per esempio una richiesta HTTP), e vogliamo evitare di bloccare l'esecuzione della nostra applicazione.